S3Client/S3Client.cs

```diff
@@ -0,0 +1,132 @@
+      };
+      _client = new AmazonS3Client(config.AccessKeyId, config.SecretAccessKey, clientConfig);
+    }
+
+    public async Task<S3Image[]> GetAlbumImages()
+    {
+      var listRequest = new ListObjectsV2Request
+      {
+        BucketName = _config.BucketName,
+        Prefix = _config.Prefix,
+      };
+
+      var response = await _client.ListObjectsV2Async(listRequest);
+      return response.S3Objects.Select(obj => new S3Image
+      {
+        Key = obj.Key,
+        Link = $"{GetBaseUrl()}/{obj.Key}",
+      }).ToArray();
+    }
+
+    public async Task<S3Image> UploadImage(string hash, string dataB64)
+    {
+      var key = $"{_config.Prefix}/{hash}";
+      var imageStream = new System.IO.MemoryStream(Convert.FromBase64String(dataB64));
+
+      // Get content type of image for S3
+      string contentType = "image/unknown";
+      using (var image = Image.FromStream(imageStream))
+      {
+        Guid guid = image.RawFormat.Guid;
+        foreach (ImageCodecInfo codec in ImageCodecInfo.GetImageDecoders())
+        {
+          if (codec.FormatID == guid)
+          {
+            contentType = codec.MimeType;
+            break;
+          }
+        }
+      }
+
+      var putRequest = new PutObjectRequest
+      {
+        BucketName = _config.BucketName,
+        Key = $"{_config.Prefix}/{hash}",
+        ContentType = contentType,
+        InputStream = imageStream,
+        CannedACL = S3CannedACL.PublicRead,
+      };
+
+      var response = await _client.PutObjectAsync(putRequest);
+      if (response.HttpStatusCode == System.Net.HttpStatusCode.OK)
+      {
+        return new S3Image
+        {
+          Key = key,
+          Link = $"{GetBaseUrl()}/{key}",
+        };
+      } else
+      {
+        return null;
+      }
+    }
+
+    public async Task<bool> TestConnection()
+    {
+      // Check if an S3 request can be successfully made
+      try
+      {
+        var listRequest = new ListObjectsV2Request
+        {
+          BucketName = _config.BucketName,
+          Prefix = _config.Prefix,
+        };
+
+        var response = await _client.ListObjectsV2Async(listRequest);
+        return response.HttpStatusCode == System.Net.HttpStatusCode.OK;
+      }
+      catch
+      {
+        return false;
+      }
+    }
+
+    public void Dispose()
+    {
+      _client?.Dispose();
+      GC.SuppressFinalize(this);
+    }
+
+    private string GetBaseUrl()
+    {
+      string url = _config.CustomDomain ?? _config.Endpoint;
+
+      // I was considering removing the url setting above entirely, as it doesn't seem to work. But it may just be my problem, so if it fails, continue to setting it through getting the bucket
+      if (string.IsNullOrEmpty(url))
+      {
+        // Fetch the bucket region
+        var regionResponse = _client.GetBucketLocationAsync(new GetBucketLocationRequest
+        {
+          BucketName = _config.BucketName
+        }).GetAwaiter().GetResult();
```

Comment:
Sounds good to me!

Reply:
Ended up not needing to use GetBucketLocation potentially, though I decided to make a private property in S3Client anyway, as I see no reason not to, and in case we decide to go back to getting bucket location, anyway.